storage/grpc_metrics.go

```diff
@@ -0,0 +1,295 @@
+	}
+	preparedResource.resource = r
+	return preparedResource, nil
+}
+
+type internalMetricsConfig struct {
+	project  string
+	endpoint string
+}
+
+type internalMetricsContext struct {
+	// monitoring API endpoint used
+	endpoint string
+	// project used by exporter
+	project string
+	// client options passed to gRPC channels
+	clientOpts []option.ClientOption
+	// instance of metric reader used by gRPC client-side metrics
+	provider *metric.MeterProvider
+	// clean func to call when closing gRPC client
+	close func()
+}
+
+func (mc *internalMetricsContext) String() string {
+	return fmt.Sprintf("endpoint: %v\nproject: %v\n", mc.endpoint, mc.project)
+}
+
+func determineMonitoringEndpoint(endpoint string) string {
+	// Check storage endpoint in case its using VPC then we use that endpoint instead.
+	if strings.Contains(endpoint, "private.googleapis.com") || strings.Contains(endpoint, "restricted.googleapis.com") {
+		return endpoint
+	}
+	// Default monitoring endpoint is used.
+	return ""
+}
+
+func createHistogramView(name, desc, unit string, boundaries []float64) metric.View {
+	return metric.NewView(metric.Instrument{
+		Name:        name,
+		Description: name,
+		Kind:        metric.InstrumentKindHistogram,
+		Unit:        unit,
+	}, metric.Stream{
+		Name:        name,
+		Description: desc,
+		Unit:        unit,
+		Aggregation: metric.AggregationExplicitBucketHistogram{Boundaries: boundaries},
+	})
+}
+
+func newGRPCMetricContext(ctx context.Context, config internalMetricsConfig) (*internalMetricsContext, error) {
+	preparedResource, err := newPreparedResource(ctx, config.project, []resource.Option{resource.WithDetectors(gcp.NewDetector())})
+	if err != nil {
+		return nil, err
+	}
+	if config.project != preparedResource.projectToUse {
+		log.Printf("The Project ID configured for metrics is %s, but the Project ID of the storage client is %s. Make sure that the service account in use has the required metric writing role (roles/monitoring.metricWriter) in the project projectIdToUse or metrics will not be written.", preparedResource.projectToUse, config.project)
+	}
+	meOpts := []mexporter.Option{
+		mexporter.WithProjectID(preparedResource.projectToUse),
+		mexporter.WithMetricDescriptorTypeFormatter(metricFormatter),
+		mexporter.WithCreateServiceTimeSeries(),
+		mexporter.WithMonitoredResourceDescription(monitoredResourceName, []string{"project_id", "location", "cloud_platform", "host_id", "instance_id", "api"})}
+	endpointToUse := determineMonitoringEndpoint(config.endpoint)
+	if endpointToUse != "" {
+		meOpts = append(meOpts, mexporter.WithMonitoringClientOptions(option.WithEndpoint(endpointToUse)))
+	}
+	exporter, err := mexporter.New(meOpts...)
+	if err != nil {
+		return nil, err
+	}
+	metricViews := []metric.View{
+		createHistogramView("grpc.client.attempt.duration", "A view of grpc.client.attempt.duration with histogram boundaries more appropriate for Google Cloud Storage RPCs", "s", latencyHistogramBoundaries()),
```

Comment:
What are these description strings used for?

Reply:
They're exposed to the user in Monitoring Metric Explorer to describe the metric. That said, it should be cached so clarifying if we really need to have this hardcoded client side.